### PR TITLE
39 continous integration building

### LINKED ISSUE #39: Continous Integration - Building
### Description

The consulting Docker image should be built continously while simultaneously checking for code quality => Linting

Requirements:
- Automatically run all lints for every PR, and do not allow merge if they do not run through
- Automatically build Docker image for `agent` that can then be tested in #38

### PR DESCRIPTION
**Pull request by timdreier**
# Description

This PR adds a GitHub action which will run the linter on PR's.

 Fixes #39 

 ## Time invested
~5h

 ## Type of change

 Please delete options that are not relevant.

 - New feature (non-breaking change which adds functionality)

 ## Does this PR introduce a breaking change?

No :)

 ## Most important changes

As described above, this change will run the linters for every PR. This is done by running the same docker containers as the precommit hook.
Additionally, the markdownlint.yml file is now mounted through the default volume to avoid the creation of a empty yml file in the root path.

 # Checklist:

 - [x] My code follows the style guidelines of this project
 - [x] I have performed a self-review of my own code
 - [x] I have commented my code, particularly in hard-to-understand areas
 - [x] I have made corresponding changes to the documentation
 - [x] My changes generate no new warnings
 - [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
 - [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

**Comment by timdreier** [comment]
@ll7 could you set the settings as described in the PR, to require the action to accept a PR?


 > You can add rules to your branches and how your merge request works.
 >
 > Your repo -> Settings -> branches -> Branch protection rules -> Add rule
 >
 > There, you will find something called Require status checks to pass before merging. Under this, you should see Status checks found in the last week for this repository. If you find the status you want to be passed before merging, you can enable the same.
 >
 > Docs: <https://help.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests>
 >
 > [(Source)](https://stackoverflow.com/questions/60776412/github-actions-is-there-a-way-to-make-it-mandatory-for-pull-request-to-merge)

**Comment by ll7** [comment]
Screenshot from 2022-11-29 14-52-27

**Comment by ll7** [comment]
I think, that this PR does not fix #39, see second requirement for building the agent.

**Comment by timdreier** [comment]
@ll7 as I understood last week, the adapted setup from @nylser already only uses one container and creates a json file with the result. So there shouldn't be a need to build a new container.

Wouldn't it be best to just use the same compose setup we use for developement in an [GitHub action](https://github.com/peter-evans/docker-compose-actions-workflow)?

**Comment by nylser** [comment]
> @ll7 as I understood last week, the adapted setup from @nylser already only uses one container and creates a json file with the result. So there shouldn't be a need to build a new container.
> 
> Wouldn't it be best to just use the same compose setup we use for developement in an [GitHub action](https://github.com/peter-evans/docker-compose-actions-workflow)?

Just chiming in, as I was mentioned: #27 will change the structure of the project a bit -- the separate `carla-ros-bridge` container won't be necessary anymore, as everything will be contained in the `agent` container.

Regardless, @ll7 is right: One of the requirements of #39 states "_creating the agent docker container_". 
This is inprecise, what I actually meant by it: Building the image of the agent Dockerfile (`docker-compose build agent`), to already have it ready for testing in the second part - #38.

I updated #39 correspondingly -- but we can discuss this tomorrow of course.

**Comment by timdreier** [comment]
@nylser thank you for summing up the new setup!

I would recommend to move the task "building the image" to #38 since it needs a physical machine in the lab. We have a storage limit of 500mb which the limit would exceed. Since we have to build the images on a physical machine anyway I don't see an advantage in building the image previously, it would just block these changes from being merged if we need to setup a physical machine as runner first.

Furthermore why should we rebuild an image? Can't we just run the stack on the physical machine?

What do you think @nylser? We can also discuss this tomorrow.

**Comment by nylser** [comment]
My thought process behind building the image: I would like to have a tagged artifact for each build. 
This way, we know for sure, that a specific tag of the image always should achieve the same results and run through. 

We can use the GitHub container registry ghcr.io to store our images, I think there almost second to none storage restrictions for public projects. Once the initial image base image is created, most of the time, only the python files will be changes, making for much smaller image deltas (Docker uses intelligent caching).

This way, our leaderboard-agent will then be a fixed artifact that can be pulled in #38 and execute it.

File: doc/02_development/09_actions.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+>
+> There, you will find something called Require status checks to pass before merging. Under this, you should see Status checks found in the last week for this repository. If you find the status you want to be passed before merging, you can enable the same.
+>
+> Docs: <https://help.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests>
+>
+> [(Source)](https://stackoverflow.com/questions/60776412/github-actions-is-there-a-way-to-make-it-mandatory-for-pull-request-to-merge)
+
+More information about creating and merging pull requests can be found [here](./08_project_management.md).
+
+## Steps
```
Comment (nylser) [inline doc/02_development/09_actions.md:67]: Do you think it is necessary to include the steps of the github actions file? I think I would rather resort to having it link to the actual .yaml file - (maybe add some comments there)?
--> To have a single source of truth

**Review by nylser — COMMENTED** [review]
The linting action already looks very good to me, but, as already discussed and to be discussed, I pledge for already building and tagging the agent image in the action.

**Comment by nylser** [comment]
As discussed, the building will be addressed in another pull request

**Review by nylser — APPROVED** [review]

**Comment by nylser** [comment]
@timdreier can you resolve the merge conflict?

File: doc/02_development/09_actions.md (hunk 1 above)
Comment (timdreier) [reply doc/02_development/09_actions.md]: Resolved :)

**Pull request merged**